common/metrics/count_metric.go

```diff
@@ -0,0 +1,100 @@
+package metrics
+
+import (
+	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var _ CountMetric = &countMetric{}
+
+// countMetric a standard implementation of the CountMetric.
+type countMetric struct {
+	Metric
+
+	// name is the name of the metric.
+	name string
+
+	// description is the description of the metric.
+	description string
+
+	// counter is the prometheus counter used to report this metric.
+	vec *prometheus.CounterVec
+
+	// labeler is the label maker used to create labels for this metric.
+	labeler *labelMaker
+}
+
+// newCountMetric creates a new CountMetric instance.
+func newCountMetric(
+	registry *prometheus.Registry,
+	namespace string,
+	name string,
+	description string,
+	labelTemplate any) (CountMetric, error) {
+
+	labeler, err := newLabelMaker(labelTemplate)
```

Comment:
The labeler becomes a no-op when the label template is nil. The purpose of using this pattern was to simplify the business logic a little. Instead of wrapping each use of the labeler in an if statement depending on whether the labeler is enabled or not, we can instead use the labeler in the same way regardless of whether or not we have a non-nil template.

This being said, if you don't like this pattern, let me know and I'll make the suggested change.